EIPS/eip-4788.md

```diff
@@ -77,76 +79,164 @@ When verifying a block, execution clients **MUST** ensure the root value in the
-parent_beacon_block_root = block_header.parent_beacon_block_root
+def get():
+    if len(evm.calldata) != 32:
+        evm.revert()
 
-sstore(HISTORY_STORAGE_ADDRESS, timestamp_index, timestamp_as_uint256)
-sstore(HISTORY_STORAGE_ADDRESS, root_index, parent_beacon_block_root)
-```
+    timestamp_idx = to_uint256_be(timestamp) % HISTORICAL_ROOTS_MODULUS
+    timestamp = storage.get(timestamp_idx)
 
-#### New stateful precompile
+    if timestamp != evm.calldata:
+        evm.revert()
 
-Beginning at the execution timestamp `FORK_TIMESTAMP`, a "stateful" precompile is deployed at `HISTORY_STORAGE_ADDRESS`.
+    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
+    root = storage.get(root_idx)
+    
+    evm.return(root)
 
-Callers of the precompile should provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
-Clients **MUST** sanitize this input call data to the precompile.
-If the input is _more_ than 32 bytes, the precompile only takes the first 32 bytes of the input buffer and ignores the rest.
-If the input is _less_ than 32 bytes, the precompile should revert.
+def set():
+    timestamp_idx = to_uint256_be(evm.timestamp) % HISTORICAL_ROOTS_MODULUS
+    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
+
+    storage.set(timestamp_idx, evm.timestamp)
+    storage.set(root_idx, evm.calldata)
+```
 
-Given this input, the precompile reduces the `timestamp` in the same way during the write routine and first checks if
-the `timestamp` recorded in the ring buffer matches the one supplied by the caller.
+##### Bytecode
+
+The exact initcode to deploy is shared below.
+
+```asm
+push1 0x58
+dup1
+push1 0x09
+push0
+codecopy
+push0
+return
+
+caller
+push20 0xfffffffffffffffffffffffffffffffffffffffe
+eq
+push1 0x44
+jumpi
+
+push1 0x20
+calldatasize
+eq
+push1 0x24
+jumpi
+
+push0
+push0
+revert
+
+jumpdest
+push3 0x018000
+push0
+calldataload
+mod
+dup1
+sload
+push0
+calldataload
+eq
+push1 0x37
+jumpi
+
+push0
+push0
+revert
+
+jumpdest
+push3 0x018000
+add
+sload
+push0
+mstore
+push1 0x20
+push0
+return
+
+jumpdest
+push3 0x018000
+timestamp
+mod
+timestamp
+dup2
+sstore
+push0
+calldataload
+swap1
+push3 0x018000
+add
+sstore
+stop
+```
 
-If the `timestamp` **does NOT** match, the client **MUST** return the "zero" word -- the 32-byte value where each byte is `0x00`.
+#### Deployment
+
+The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated
+by working backwards from the desired deployment transaction:
+
+```json
+{
+  "type": "0x0",
+  "nonce": "0x0",
+  "to": null,
+  "gas": "0x27eac",
+  "gasPrice": "0xe8d4a51000",
+  "maxPriorityFeePerGas": null,
+  "maxFeePerGas": null,
+  "value": "0x0",
+  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
+  "v": "0x1b",
+  "r": "0x539",
+  "s": "0x1337",
```

Comment:
Alternative suggestion: 
```suggestion
  "s": "0x133700cb4fbb1368012b31",
```

Then we get a `bEaC0` -prefixed address `0xbEaC0B09a6ed668e6BC5d29e29013aCb9286bF5F` with the transaction
```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",
  "s": "0x133700cb4fbb1368012b31",
  "hash": "0xa71141ed853de6fdd7c9d9355293f255011e9e512cb833a03b3ff20d8045f5b9"
}
```

Reply:
Update: `BeAC00` - prefix:

Address: `0xBeAC00541d49391ED88ABF392bfC1F4dEa8c4143`
Tx:
```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",
  "s": "0x1337007db5d301ac899c9e",
  "hash": "0x6923750e3b16ef9bfe1a70820d0d0628d7d4e84906b19ef91840fe913d688759"
}
```

Reply:
amazing!

Reply:
Ah wait did you intend for this to be the sender of the transaction or the address of the deployed contract? Because I am recovering `0xBeAC00541d49391ED88ABF392bfC1F4dEa8c4143` as the EOA address and `0xC221266EE238DbA34c6df82f75516a27087C0Bd9` as the deployed address.

Reply:
Ah silly me. Let me get back to you on this

Reply:
How about this then? 
_deploy_ address: `0xbEac00dDB15f3B6d645C48263dC93862413A222D`

```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",
  "s": "0x133700f3a77843802897db",
  "hash": "0x14789a20c0508b81ab7a0287a12a3a41ca960aa18244af8e98689e37ed569f07"
}
```